rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -57,44 +59,82 @@ def replication_enabled(self):
             logger.info(
                 "[Dual Write] Generate relations from current state of role(%s): '%s'", self.role.uuid, self.role.name
             )
-            relations = migrate_role(
-                self.role, False, str(self.root_workspace.uuid), self.org_id, True, True, True, False
+
+            self.binding_mapping = self.role.binding_mapping
+
+            if self.binding_mapping is None:
+                raise Exception(f"BindingMapping for role {self.role.uuid} not found.")
+
+            relations, _ = migrate_role(
+                self.role,
+                write_relationships=False,
+                root_workspace=str(self.root_workspace.uuid),
+                default_workspace=self.org_id,
+                current_bindings=self.binding_mapping,
             )
+
             self.current_role_relations = relations
         except Exception as e:
             raise DualWriteException(e)
 
-    def regenerate_relations_and_mappings_for_role(self):
-        """Delete and generated relations with mapping for a role."""
+    def generate_replication_event_to_outbox(self, role):
+        """Generate replication event to outbox table."""
         if not self.replication_enabled():
-            return []
-        return self.generate_relations_and_mappings_for_role()
+            return
+        self.role = role
+        self._generate_relations_and_mappings_for_role()
+        return self.save_replication_event_to_outbox()
+
+    def save_replication_event_to_outbox(self):
+        """Generate and store replication event to outbox table."""
+        if not self.replication_enabled():
+            return {}
+        try:
+            replication_event = self._build_replication_event()
+            self._save_replication_event(replication_event)
+        except Exception as e:
+            raise DualWriteException(e)
+        return replication_event
 
-    def generate_relations_and_mappings_for_role(self):
+    def _generate_relations_and_mappings_for_role(self):
         """Generate relations and mappings for a role with new UUIDs for v2 role and role bindings."""
         if not self.replication_enabled():
             return []
         try:
             logger.info("[Dual Write] Generate new relations from role(%s): '%s'", self.role.uuid, self.role.name)
-            relations = migrate_role(self.role, False, str(self.root_workspace.uuid), self.org_id, True)
+
+            relations, mappings = migrate_role(
+                self.role,
+                write_relationships=False,
+                root_workspace=str(self.root_workspace.uuid),
+                default_workspace=self.org_id,
+                current_bindings=self.binding_mapping)
+
             self.role_relations = relations
+
+            if self.binding_mapping is None:
+                self.binding_mapping = BindingMapping.objects.create(role=self.role)
+
+            self.binding_mapping.mappings = mappings
+            self.binding_mapping.save()
+
```

Comment:
I moved the mapping update and persist out of `migrate_role` to (at least for me) make the logic clearer and simplify `migrate_role` arguments and implementation. All mapping SQL is now performed inside the dual write handler.